DAQ/src/ArtBinaryPacketsFromDigis_module.cc

```diff
@@ -0,0 +1,1113 @@
+    void   fillCalorimeterDataPacket(const CaloDigi& SD, CaloDataPacket& caloData);
+    void   addCaloHitToCaloPacket(calo_data_block_t& dataBlock,
+				  CaloDataPacket& caloData);
+
+    void   fillCalorimeterHeaderDataPacket(const CaloDigi& SD, DataBlockHeader& HeaderData, uint64_t& EventNum);
+
+    void   fillEmptyCalorimeterDataPacket(CaloDataPacket& caloData);
+    void   fillHeaderByteAndPacketCounts(calo_data_block_t& caloData);
+
+    void   processCalorimeterData(art::Event& evt, uint64_t& eventNum,
+				  calo_data_block_list_t& caloDataBlocks);
+
+
+    void   fillCalorimeterDMABlocks(raw_data_list_t& dataStream, calo_data_block_list_t& caloData);
+
+    void   fillCalorimeterDataStream(raw_data_list_t& dataStream, calo_data_block_t& caloData);
+
+    void   printCalorimeterData(CaloDataPacket const& curDataBlock);
+
+    size_t waveformMaximumIndex(std::vector<adc_t>& waveform);
+
+    void flushBuffer(raw_data_list_t& buffers);
+    void closeDataBuffer(raw_data_list_t& dataStream, bool openNew = true) {     
+      uint64_t sz = dataStream.back().second;
+      if (_diagLevel > 3) {
+	std::cout<< "[closeDataBuffer] 1) sz = "<<sz<<std::endl;
+	std::cout << "[closeDataBuffer] 1a) sz = "<< dataStream.back().second << std::endl;
+      }
+      if (sz < 48) {
+	if (_diagLevel > 3) {
+	  std::cout<< "[closeDataBuffer] dataStream.back().first = "<<dataStream.back().first<<std::endl;
+	  std::cout<< "[closeDataBuffer] dataStream.back().first[sz] = "<<dataStream.back().first[sz]<<std::endl;
+	  std::cout<< "[closeDataBuffer] &dataStream.back().first[sz] = "<<&dataStream.back().first[sz]<<std::endl;
+	}
+	bzero(&dataStream.back().first[sz], 48 - sz);
+	sz = 48;
+      }
+      if (_diagLevel > 3) {
+	std::cout << "[closeDataBuffer] dataStream.size = "<< dataStream.size() << std::endl;
+	if (dataStream.size()> 0){
+	  std::cout <<  "[closeDataBuffer] dataStream.back().second = " << dataStream.back().second <<std::endl;
+	}
+	std::cout<< "[closeDataBuffer] 2) sz = "<<sz<<std::endl;
+      }
+      uint64_t ex_sz = sz - 16;
+
+      memcpy(&dataStream.back().first, &sz, sizeof(uint64_t));
+      memcpy(&dataStream.back().first[8], &ex_sz, sizeof(uint64_t));
+
+      if (openNew) {
+	dataStream.push_back(std::pair<mu2e_databuff_t, size_t>());
+	dataStream.back().second = 16; // DMA Size words
+      }
+    }
+
+    std::vector<adc_t> generateDMABlockHeader(size_t theCount) const;
+    std::vector<adc_t> generateEventByteHeader(size_t theCount) const;
+  };
+
+  //--------------------------------------------------------------------------------
+  // temporary function used to find the location of the waveform peak in the 
+  // calorimeter digitized waveform
+  //--------------------------------------------------------------------------------
+  size_t ArtBinaryPacketsFromDigis::waveformMaximumIndex(std::vector<adc_t>& waveform) {
+    size_t  indexMax(0), content(0);
```

Comment:
1) The argument should be const& and the member function should be const; please check for this everywhere.

2) I think that there is a standard library function that will do this for you.  I would implement it using that.  If there is a standard library function to do what you want, the best you can do is equal it - it's highly unlikely you can do better.  Not quite the second law of thermodynamics, but close.

3) This does not need to be a member function. If something does not need to be a member function, make it a free function.  It can live in its own file if it is general interest; if it is so specialized that it really only belongs in this file, put it in an anonymous namespace at the top of the file.

Reply:
Hi Rob,

Once we will start using Sophie's improved caloDigi, I will delete this function, because the peak info will be available from it. So, if you agree, I will just implement the conversion to const and const&.

Reply:
OK.  Please go ahead with that plan.

As you move forward, my advice is that if functions do not need access to member data, make them free functions.    The reason is that you don't want to give a function permission to modify something that it does not need to modify.